sdks/python/apache_beam/runners/interactive/pipeline_graph.py

```diff
@@ -70,9 +95,25 @@ def __init__(self,
   def get_dot(self):
     return str(self._get_graph())
 
-  def _is_top_level_transform(self, transform):
-    return transform.unique_name and '/' not in transform.unique_name \
-        and not transform.unique_name.startswith('ref_')
+  def display_graph(self):
+    """Displays graph via IPython or prints DOT if not possible."""
+    try:
+      from IPython.core import display  # pylint: disable=import-error
+      display.display(display.HTML(self._get_graph().create_svg()))  # pylint: disable=protected-access
+    except ImportError:
+      print(str(self._get_graph()))
+
+  def _top_level_transforms(self):
+    """Yields all top level PTransforms (subtransforms of the root PTransform).
+
+    Yields: (str, PTransform proto) ID, proto pair of top level PTransforms.
+    """
+    transforms = self._pipeline_proto.components.transforms
+    for root_transform_id in self._pipeline_proto.root_transform_ids:
+      root_transform_proto = transforms[root_transform_id]
+      for top_level_transform_id in root_transform_proto.subtransforms:
+        top_level_transform_proto = transforms[top_level_transform_id]
+        yield top_level_transform_id, top_level_transform_proto
```

Comment:
I am not sure that I understand correctly. What is a 'top level transform'? And what is a root transform?

If I understand correctly, let's consider a pipeline that contains a reshuffle. Something like so: `Read -> Reshuffle -> Write`, and the reshuffle transform looks like so: [image](https://user-images.githubusercontent.com/1301740/44279944-20e33780-a208-11e8-84b7-722de852e988.png).

Transforms in the pipeline are then:
```
Read
Reshuffle
Reshuffle/AddRandomKeys
Reshuffle/ReshufflePerKey/
Reshuffle/ReshufflePerKey/Map
Reshuffle/ReshufflePerKey/GroupByKey
Reshuffle/ReshufflePerKey/FlatMap
Reshuffle/RemoveRandomKeys
Write
```

The root transforms would be `Read`, `Reshuffle`, `Write`. Is this correct? - or maybe these are the top level transforms? : ) Maybe the root transform is just `/` from which all come later?

Reply:
Root transforms are where the pipeline execution starts. There is a [field in Pipeline proto](https://github.com/apache/beam/blob/53d57bf642a03d611909015249f3f9b94675b2a2/model/pipeline/src/main/proto/beam_runner_api.proto#L100) that points to such PTransform(s). In the case of your example, the root transform is a transform with Read, Reshuffle, and Write as it subtransforms, and yes it is something like `/` that include everything else.

Top level transforms will PTransforms applied (and usually assigned a unique label) **by the user**. In this case, Read, Reshuffle, and Write are top level transforms.

Reply:
ah I see.. and we only want to show user-named transforms in the rendered graph, right?